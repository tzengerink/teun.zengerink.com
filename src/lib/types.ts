export enum Width {
  Mobile = 768,
  Desktop = 1280,
}

export interface ConfigCaption {
  key: string
  caption: string
}

export interface ConfigItem {
  title: string
  statement?: string
  isArchived?: boolean
  captions?: ConfigCaption[]
}

export interface Export {
  width: Width
  url: string
}

export interface Photo {
  key: string
  caption?: string
  exports: Export[]
}

export interface Project {
  title: string
  slug: string
  photos: Photo[]
  statement?: string
  isArchived?: boolean
}

export const MAX_WIDTHS = [Width.Mobile, Width.Desktop]
