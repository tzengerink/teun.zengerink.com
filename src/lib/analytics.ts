export const trackPageView = (): void => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  const method = process.env.NEXT_PUBLIC_PIWIK_APP_ID ? _paq.push : console.info
  method(['trackPageView'])
}
