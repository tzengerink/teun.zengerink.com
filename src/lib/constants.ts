export const TITLE_ROOT = 'Teun Zengerink'
export const TITLE_DIV = '-'
