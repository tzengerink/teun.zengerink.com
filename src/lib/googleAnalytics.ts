declare global {
  interface Window {
    gtag: (...any) => void
  }
}

export const trackPageView = (url: string): void => {
  if (!process.env.NEXT_PUBLIC_TRACKING_ID) return console.info(`[Tracking Pageview] ${url}`)
  window.gtag('config', process.env.NEXT_PUBLIC_TRACKING_ID, {
    page_path: url,
  })
}
