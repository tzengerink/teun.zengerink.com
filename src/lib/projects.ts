import fs from 'fs'
import path from 'path'
import slugify from 'slugify'
import yaml from 'js-yaml'

export enum Width {
  Mobile = 768,
  Desktop = 1280,
}

interface ConfigCaption {
  key: string
  caption: string
}

interface ConfigItem {
  title: string
  statement?: string
  isArchived?: boolean
  captions?: ConfigCaption[]
}

export interface Export {
  width: Width
  url: string
}

export interface Photo {
  key: string
  caption?: string
  exports: Export[]
}
export interface Project {
  title: string
  slug: string
  photos: Photo[]
  statement?: string
  isArchived?: boolean
}

export const MAX_WIDTHS = [Width.Mobile, Width.Desktop]

const generatePhotos = (slug: string, captions: ConfigCaption[]): Photo[] => {
  const photosDirectory = path.join(process.cwd(), 'public/photos', slug, `${Width.Mobile}w`)
  const fileNames = fs.readdirSync(photosDirectory)

  return fileNames.map((filename) => {
    const key = path.basename(filename, '.jpg')
    const caption = captions?.find((c) => c.key == key)

    return {
      key,
      caption: caption?.caption ?? '',
      exports: MAX_WIDTHS.map((width) => ({ width, url: `/photos/${slug}/${width}w/${filename}` })),
    }
  })
}

const generateProjects = (item: ConfigItem[]): Project[] => {
  return item.map((project) => {
    const slug = slugify(project.title, { lower: true })

    return {
      slug,
      title: project.title,
      statement: project.statement ?? '',
      isArchived: project.isArchived ?? false,
      photos: generatePhotos(slug, project.captions),
    }
  })
}

export const getProjects = async (): Promise<Project[]> => {
  let projects: Project[] = []

  try {
    const file = path.join(process.cwd(), 'src', 'config.yml')
    const config = yaml.load(fs.readFileSync(file, 'utf-8'))
    projects = await generateProjects(config)
  } catch (e) {
    console.error(e)
  }

  return projects
}
