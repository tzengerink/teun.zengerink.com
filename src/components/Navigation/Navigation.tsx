'use client'

import classNames from 'classnames'
import React, { useState } from 'react'
import Link from 'next/link'

import { Project } from '../../lib/projects'
import Hamburger from '../Hamburger/Hamburger'
import Menu from '../Menu/Menu'

interface NavigationProps {
  projects: Project[]
  pageTitle: string
}

const Navigation: React.FC<NavigationProps> = ({ pageTitle, projects }): React.ReactElement => {
  const [isOpen, setIsOpen] = useState(false)
  const filteredProjects = projects.filter((project) => !project.isArchived)

  return (
    <aside>
      <Link href="/">
        <h1
          className={classNames(
            'fixed',
            'z-40',
            'top-[6px]',
            'left-md',
            'pt-4',
            'font-medium',
            'uppercase',
            'hover:underline',
          )}
        >
          {pageTitle}
        </h1>
      </Link>
      <Hamburger isOpen={isOpen} onClick={() => setIsOpen(!isOpen)} />
      <Menu
        projects={filteredProjects}
        isOpen={isOpen}
        onOpen={() => setIsOpen(!isOpen)}
        onItemClick={() => setIsOpen(false)}
      />
      <Link
        className={classNames(
          'absolute',
          'z-20',
          'bottom-md',
          'right-md',
          'text-xs',
          'text-grey/80',
          'uppercase',
          'hover:underline',
        )}
        href="https://www.instagram.com/t.zengerink/"
      >
        Instagram
      </Link>
    </aside>
  )
}

export default Navigation
