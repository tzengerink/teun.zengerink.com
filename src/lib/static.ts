import { createGetProjects, Project } from './projects'

interface Path {
  params: { slug: string[] }
}

export interface Props {
  projects: Project[]
}

const getProjects = createGetProjects()

export const getStaticPaths = async (): Promise<{ paths: Path[]; fallback: boolean }> => {
  const projects = await getProjects()

  return {
    paths: projects.flatMap((project) => {
      const paths = project.photos.map((photo) => ({ params: { slug: [project.slug, photo.key] } }))
      if (project.statement) paths.unshift({ params: { slug: [project.slug] } })
      return paths
    }),
    fallback: false,
  }
}

export const getStaticProps = async (): Promise<{ props: Props }> => ({ props: { projects: await getProjects() } })
