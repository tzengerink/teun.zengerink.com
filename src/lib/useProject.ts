import { Photo, Project } from './projects'
import { useRouter } from 'next/router'

interface UseProject {
  activeKey: string | undefined
  previous: () => void
  next: () => void
}

const NOOP = () => false

export const useProject = (project: Project): UseProject => {
  if (!project?.photos?.length) return { activeKey: undefined, previous: NOOP, next: NOOP }

  const router = useRouter()

  const activeKey = router?.query?.slug?.length > 1 ? router.query?.slug[1] : project.photos[0].key
  const slideIndex = project.photos.findIndex((photo) => photo.key === activeKey)
  const firstPhoto = project.photos[0]
  const lastPhoto = [...project.photos].pop()
  const nextPhoto = project.photos[slideIndex + 1]
  const previousPhoto = project.photos[slideIndex - 1]
  const isLast = activeKey === lastPhoto.key
  const isFirst = activeKey === firstPhoto.key

  const go = (photo: Photo) => router.push(`${`/work/${project.slug}`}/${photo.key}`)
  const previous = () => (isFirst ? go(lastPhoto) : go(previousPhoto))
  const next = () => (isLast ? go(firstPhoto) : go(nextPhoto))

  return { activeKey, previous, next }
}
