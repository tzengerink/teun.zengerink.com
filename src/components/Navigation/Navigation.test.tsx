import React from 'react'
import { fireEvent, render } from '@testing-library/react'
import mockProjects from '../../__mocks__/projects'
import Navigation from './Navigation'
import { Project } from '../../lib/types'

const mockRouter = jest.fn(() => ({ asPath: '/' }))
jest.mock('next/router', () => ({ useRouter: () => mockRouter() }))

jest.mock('../Menu/Menu', () => {
  const Mock = ({ projects, isOpen, onItemClick }) => (
    <button onClick={onItemClick}>
      <div>{isOpen ? '[Menu open]' : '[Menu closed]'}</div>
      <ul>
        {projects.map((project: Project) => (
          <li key={project.slug}>{project.isArchived ? '[Archived project]' : '[Project]'}</li>
        ))}
      </ul>
    </button>
  )
  Mock.displayName = 'Menu'
  return Mock
})

jest.mock('../Hamburger/Hamburger', () => {
  const Mock = ({ isOpen, onClick }) => (
    <button onClick={onClick}>{isOpen ? '[Hamburger open]' : '[Hamburger closed]'}</button>
  )
  Mock.displayName = 'Hamburger'
  return Mock
})

const defaultProps = { pageTitle: 'My Homepage', projects: mockProjects }

beforeEach(jest.clearAllMocks)

describe('Navigation', () => {
  const renderComponent = (props = defaultProps) => render(<Navigation {...props} />)

  it('renders correctly', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })

  it('filters out archived projects', () => {
    const { queryByText } = renderComponent()
    const found = queryByText('[Archived project]')
    expect(found).toBeNull()
  })

  it('toggles the menu when the hamburger is clicked', () => {
    const { getByText } = renderComponent()
    expect(getByText('[Menu closed]')).toBeDefined()
    fireEvent.click(getByText('[Hamburger closed]'))
    expect(getByText('[Menu open]')).toBeDefined()
    fireEvent.click(getByText('[Hamburger open]'))
    expect(getByText('[Menu closed]')).toBeDefined()
  })

  it('closes the menu when an item is clicked', () => {
    const { getByText } = renderComponent()
    fireEvent.click(getByText('[Hamburger closed]'))
    expect(getByText('[Menu open]')).toBeDefined()
    fireEvent.click(getByText('[Menu open]'))
    expect(getByText('[Menu closed]')).toBeDefined()
  })
})
