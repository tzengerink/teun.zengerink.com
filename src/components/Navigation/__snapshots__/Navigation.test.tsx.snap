// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation renders correctly 1`] = `
<div>
  <aside>
    <a
      href="/"
    >
      <h1
        class="fixed z-40 top-[6px] left-md pt-4 font-medium uppercase hover:underline"
      >
        My Homepage
      </h1>
    </a>
    <button>
      [Hamburger closed]
    </button>
    <button>
      <div>
        [Menu closed]
      </div>
      <ul>
        <li>
          [Project]
        </li>
        <li>
          [Project]
        </li>
      </ul>
    </button>
    <a
      class="absolute z-20 bottom-md right-md text-xs text-grey/80 uppercase hover:underline"
      href="https://www.instagram.com/t.zengerink/"
    >
      Instagram
    </a>
  </aside>
</div>
`;
